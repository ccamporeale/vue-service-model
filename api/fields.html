<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Fields API | vue-service-model</title>
    <meta name="generator" content="VuePress 1.5.0">
    <link rel="icon" href="/vue-service-model/logo.png">
    <meta name="description" content="Vue.js library for handling REST service requests with caching, aggregation and model definitions">
    <link rel="preload" href="/vue-service-model/assets/css/0.styles.a218223d.css" as="style"><link rel="preload" href="/vue-service-model/assets/js/app.c2b3a9a1.js" as="script"><link rel="preload" href="/vue-service-model/assets/js/3.45671bfb.js" as="script"><link rel="preload" href="/vue-service-model/assets/js/8.64e7feff.js" as="script"><link rel="prefetch" href="/vue-service-model/assets/js/10.22689168.js"><link rel="prefetch" href="/vue-service-model/assets/js/11.ba383af0.js"><link rel="prefetch" href="/vue-service-model/assets/js/12.375d0476.js"><link rel="prefetch" href="/vue-service-model/assets/js/13.804a0c41.js"><link rel="prefetch" href="/vue-service-model/assets/js/14.d6887d0c.js"><link rel="prefetch" href="/vue-service-model/assets/js/15.f93a33c2.js"><link rel="prefetch" href="/vue-service-model/assets/js/16.568206dd.js"><link rel="prefetch" href="/vue-service-model/assets/js/17.c3d9d7ac.js"><link rel="prefetch" href="/vue-service-model/assets/js/18.247cde34.js"><link rel="prefetch" href="/vue-service-model/assets/js/19.1d1f842e.js"><link rel="prefetch" href="/vue-service-model/assets/js/20.1cb7654d.js"><link rel="prefetch" href="/vue-service-model/assets/js/21.7db86893.js"><link rel="prefetch" href="/vue-service-model/assets/js/22.e933c31d.js"><link rel="prefetch" href="/vue-service-model/assets/js/4.973410df.js"><link rel="prefetch" href="/vue-service-model/assets/js/5.c2a7c680.js"><link rel="prefetch" href="/vue-service-model/assets/js/6.b5c2df69.js"><link rel="prefetch" href="/vue-service-model/assets/js/7.612f3f77.js"><link rel="prefetch" href="/vue-service-model/assets/js/9.4a145423.js"><link rel="prefetch" href="/vue-service-model/assets/js/vendors~docsearch.a7fb83ab.js">
    <link rel="stylesheet" href="/vue-service-model/assets/css/0.styles.a218223d.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container"><header class="navbar"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/vue-service-model/" class="home-link router-link-active"><img src="/vue-service-model/logo.png" alt="vue-service-model" class="logo"> <span class="site-name can-hide">vue-service-model</span></a> <div class="links"><form id="search-form" role="search" class="algolia-search-wrapper search-box"><input id="algolia-search-input" class="search-query"></form> <nav class="nav-links can-hide"><div class="nav-item"><a href="/vue-service-model/guide/" class="nav-link">
  Guide
</a></div><div class="nav-item"><a href="/vue-service-model/api/" class="nav-link router-link-active">
  API Reference
</a></div><div class="nav-item"><a href="https://github.com/freakzlike/vue-service-model/releases" target="_blank" rel="noopener noreferrer" class="nav-link external">
  Release Notes
  <svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></div> <a href="https://github.com/freakzlike/vue-service-model" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></nav></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><nav class="nav-links"><div class="nav-item"><a href="/vue-service-model/guide/" class="nav-link">
  Guide
</a></div><div class="nav-item"><a href="/vue-service-model/api/" class="nav-link router-link-active">
  API Reference
</a></div><div class="nav-item"><a href="https://github.com/freakzlike/vue-service-model/releases" target="_blank" rel="noopener noreferrer" class="nav-link external">
  Release Notes
  <svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></div> <a href="https://github.com/freakzlike/vue-service-model" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></nav>  <ul class="sidebar-links"><li><a href="/vue-service-model/" class="sidebar-link">Home</a></li><li><section class="sidebar-group depth-0"><a href="/vue-service-model/guide/" class="sidebar-heading clickable"><span>Guide</span> <!----></a> <ul class="sidebar-links sidebar-group-items"><li><a href="/vue-service-model/guide/getting-started.html" class="sidebar-link">Getting started</a></li><li><a href="/vue-service-model/guide/installation.html" class="sidebar-link">Installation</a></li><li><a href="/vue-service-model/guide/models.html" class="sidebar-link">Models</a></li><li><a href="/vue-service-model/guide/fields.html" class="sidebar-link">Fields</a></li><li><a href="/vue-service-model/guide/field-types.html" class="sidebar-link">Field types</a></li><li><a href="/vue-service-model/guide/service-model.html" class="sidebar-link">ServiceModel</a></li><li><a href="/vue-service-model/guide/model-manager.html" class="sidebar-link">ModelManager</a></li><li><a href="/vue-service-model/guide/components.html" class="sidebar-link">Components</a></li><li><a href="/vue-service-model/guide/configuration.html" class="sidebar-link">Configuration</a></li></ul></section></li><li><section class="sidebar-group depth-0"><p class="sidebar-heading open"><span>API</span> <!----></p> <ul class="sidebar-links sidebar-group-items"><li><a href="/vue-service-model/api/base-model.html" class="sidebar-link">BaseModel API</a></li><li><a href="/vue-service-model/api/service-model.html" class="sidebar-link">ServiceModel API</a></li><li><a href="/vue-service-model/api/fields.html" class="active sidebar-link">Fields API</a></li><li><a href="/vue-service-model/api/model-manager.html" class="sidebar-link">ModelManager API</a></li><li><a href="/vue-service-model/api/configuration.html" class="sidebar-link">Configuration API</a></li></ul></section></li></ul> </aside> <main class="page"> <div class="theme-default-content content__default"><h1 id="fields-api"><a href="#fields-api" class="header-anchor">#</a> Fields API</h1> <div class="language-typescript extra-class"><pre class="language-typescript"><code><span class="token keyword">class</span> <span class="token class-name">Field</span> <span class="token punctuation">{</span>
  <span class="token comment">// Constructor takes field definition</span>
  <span class="token keyword">constructor</span> <span class="token punctuation">(</span>def<span class="token operator">:</span> FieldDef<span class="token punctuation">)</span>

  <span class="token comment">// Clone field instance</span>
  <span class="token keyword">public</span> <span class="token function">clone</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> Field

  <span class="token comment">// Returns field name (Which has been set as key at fieldsDef.</span>
  <span class="token comment">// Will throw FieldNotBoundException in case field has not been bound to a model</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">name</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">string</span>

  <span class="token comment">// Returns either attributeName from fieldsDef or field name</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">attributeName</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">string</span>

  <span class="token comment">// Returns field definition</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">definition</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> FieldDef

  <span class="token comment">// Assigned model</span>
  <span class="token comment">// Will throw FieldNotBoundException in case field has not been bound to a model</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">model</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> BaseModel

  <span class="token comment">// Returns async field value from data by calling valueGetter with data of assigned model</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">value</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">any</span>

  <span class="token comment">// Sets field value to model data by calling valueSetter with data of assigned model</span>
  <span class="token keyword">public</span> <span class="token keyword">set</span> <span class="token function">value</span> <span class="token punctuation">(</span>value<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">)</span>

  <span class="token comment">// Returns async field label from field definition</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">label</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span><span class="token builtin">string</span><span class="token operator">&gt;</span>

  <span class="token comment">// Returns async field hint from field definition</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">hint</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span><span class="token builtin">string</span><span class="token operator">&gt;</span>

  <span class="token comment">// Returns boolean whether field is a primary key</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">isPrimaryKey</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">boolean</span>

  <span class="token comment">// Retrieve value from data structure according to attributeName</span>
  <span class="token comment">// Uses nested syntax from attributeName (e.g. &quot;address.city&quot; -&gt; {address: {city: 'New York'}})</span>
  <span class="token comment">// Will return null if value is not available</span>
  <span class="token keyword">public</span> <span class="token function">valueGetter</span> <span class="token punctuation">(</span>data<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">any</span>

  <span class="token comment">// Set value to data by using attributeName</span>
  <span class="token comment">// Will create nested structure from attributeName (e.g. &quot;address.city&quot; -&gt; {address: {city: 'New York'}})</span>
  <span class="token keyword">public</span> <span class="token function">valueSetter</span> <span class="token punctuation">(</span>value<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">,</span> data<span class="token operator">:</span> Dictionary<span class="token operator">&lt;</span><span class="token builtin">any</span><span class="token operator">&gt;</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token keyword">void</span>

  <span class="token comment">// Display component to render when displaying value with &lt;display-field/&gt;</span>
  <span class="token comment">// For more information see Field - Rendering </span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">displayComponent</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span>ComponentModule<span class="token operator">&gt;</span>

  <span class="token comment">// Async function to prepare before displayRender gets called</span>
  <span class="token comment">// Can return any data which needs to be resolved for displayRender</span>
  <span class="token keyword">public</span> <span class="token keyword">async</span> <span class="token function">prepareDisplayRender</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span><span class="token builtin">any</span><span class="token operator">&gt;</span>

  <span class="token comment">// Simple Vue render function when using default displayComponent when displaying value with &lt;display-field/&gt;</span>
  <span class="token comment">// For more information see Field - Rendering </span>
  <span class="token keyword">public</span> <span class="token function">displayRender</span> <span class="token punctuation">(</span>h<span class="token operator">:</span> CreateElement<span class="token punctuation">,</span> renderData<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">)</span><span class="token operator">:</span> VNode

  <span class="token comment">// Input component to render when showing input for field with &lt;input-field/&gt;</span>
  <span class="token comment">// For more information see Field - Rendering </span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">inputComponent</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span>ComponentModule<span class="token operator">&gt;</span>

  <span class="token comment">// Async function to prepare before inputRender gets called</span>
  <span class="token comment">// Can return any data which needs to be resolved for inputRender</span>
  <span class="token keyword">public</span> <span class="token keyword">async</span> <span class="token function">prepareInputRender</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span><span class="token builtin">any</span><span class="token operator">&gt;</span>

  <span class="token comment">// Simple Vue render function when using default inputComponent for input of field value with &lt;input-field/&gt;</span>
  <span class="token comment">// For more information see Field - Rendering </span>
  <span class="token keyword">public</span> <span class="token function">inputRender</span> <span class="token punctuation">(</span>h<span class="token operator">:</span> CreateElement<span class="token punctuation">,</span> renderData<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">)</span><span class="token operator">:</span> VNode

<span class="token punctuation">}</span>
</code></pre></div></div> <footer class="page-edit"><div class="edit-link"><a href="https://github.com/freakzlike/vue-service-model/edit/master/docs/api/fields.md" target="_blank" rel="noopener noreferrer">Edit this page on GitHub</a> <svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></div> <div class="last-updated"><span class="prefix">Last Updated:</span> <span class="time">5/3/2020, 8:56:57 AM</span></div></footer> <div class="page-nav"><p class="inner"><span class="prev">
      ←
      <a href="/vue-service-model/api/service-model.html" class="prev">
        ServiceModel API
      </a></span> <span class="next"><a href="/vue-service-model/api/model-manager.html">
        ModelManager API
      </a>
      →
    </span></p></div> </main></div><div class="global-ui"></div></div>
    <script src="/vue-service-model/assets/js/app.c2b3a9a1.js" defer></script><script src="/vue-service-model/assets/js/3.45671bfb.js" defer></script><script src="/vue-service-model/assets/js/8.64e7feff.js" defer></script>
  </body>
</html>
