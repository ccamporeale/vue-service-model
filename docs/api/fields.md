# Fields API

```typescript
class Field {
  // Constructor takes field definition
  constructor (def: FieldDef)

  // Clone field instance
  public clone (): Field

  // Returns field name (Which has been set as key at fieldsDef.
  // Will throw FieldNotBoundException in case field has not been bound to a model
  public get name (): string

  // Returns either attributeName from fieldsDef or field name
  public get attributeName (): string

  // Returns field definition
  public get definition (): FieldDef

  // Assigned model
  // Will throw FieldNotBoundException in case field has not been bound to a model
  public get model (): BaseModel

  // Returns async field value from data by calling valueGetter with data of assigned model
  public get value (): any

  // Sets field value to model data by calling valueSetter with data of assigned model
  public set value (value: any)

  // Returns async field label from field definition
  public get label (): Promise<string>

  // Returns async field hint from field definition
  public get hint (): Promise<string>

  // Returns async field options with validation and default values depending on field type
  public get options (): Promise<FieldTypeOptions>

  // Validate field options and set default values depending on field type
  protected async validateOptions (options: FieldTypeOptions): Promise<FieldTypeOptions>

  // Returns boolean whether field is a primary key
  public get isPrimaryKey (): boolean

  // Retrieve value from data structure according to attributeName
  // Uses nested syntax from attributeName (e.g. "address.city" -> {address: {city: 'New York'}})
  // Will return null if value is not available
  public valueGetter (data: any): any

  // Set value to data by using attributeName
  // Will create nested structure from attributeName (e.g. "address.city" -> {address: {city: 'New York'}})
  public valueSetter (value: any, data: Dictionary<any>): void

  // Display component to render when displaying value with <display-field/>
  // For more information see Field - Rendering 
  public get displayComponent (): Promise<ComponentModule>

  // Async function to prepare before displayRender gets called
  // Can return any data which needs to be resolved for displayRender
  public async prepareDisplayRender (): Promise<any>

  // Simple Vue render function when using default displayComponent when displaying value with <display-field/>
  // For more information see Field - Rendering 
  public displayRender (h: CreateElement, renderData: any): VNode

  // Input component to render when showing input for field with <input-field/>
  // For more information see Field - Rendering 
  public get inputComponent (): Promise<ComponentModule>

  // Async function to prepare before inputRender gets called
  // Can return any data which needs to be resolved for inputRender
  public async prepareInputRender (): Promise<any>

  // Simple Vue render function when using default inputComponent for input of field value with <input-field/>
  // For more information see Field - Rendering 
  public inputRender (h: CreateElement, renderData: any): VNode

}
```
