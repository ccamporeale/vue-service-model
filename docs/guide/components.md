# Components

[[toc]]

## DisplayField

The `DisplayField` component can be used to display the value of the given field. The property `model` can either be a 
model instance or `null` to allow async loading of the model instance. The component will not render when `null` has 
been passed as `model`. To change the output for specific fields see [Fields rendering](/guide/fields.html#rendering). 

```vue
<template>
  [...]
    <display-field :model="album" field-name="title"/>
    <!-- or directly with field -->
    <display-field :field="album.getField('title')"/>
  [...]
</template>

<script>
  import {DisplayField} from 'vue-service-model'

  export default {
    components: {DisplayField},
    data () {
      return {
        album: null
      }  
    },
    created () {
      this.loadAlbum()    
    },
    methods: {
      async loadAlbum () {
        this.album = await Album.objects.detail(1)
      }
    }
  }
</script>
```

::: warning
When using direct `field` property, keep in mind that the model will be `null` during loading
:::

To display the loading state of the component, when the data is being fetched, you can use the `loading` slot.

```vue
<template>
  [...]
    <display-field :model="album" field-name="title">
      <template v-slot:loading>
        <span>Loading title...</span>
      </template>
    </display-field>
  [...]
</template>
```

## InputField

The `InputField` component is equal to the `DisplayField`. The input value will directly change the data of your model (using the `valueSetter`). 

```vue
<template>
  [...]
    <input-field :model="album" field-name="title"/>
    <!-- or directly with field -->
    <input-field :field="album.getField('title')"/>
  [...]
</template>

<script>
  import {InputField} from 'vue-service-model'

  export default {
    components: {InputField},
    data () {
      return {
        album: null
      }  
    },
    created () {
      this.loadAlbum()    
    },
    methods: {
      async loadAlbum () {
        this.album = await Album.objects.detail(1)
      }
    }
  }
</script>
```

::: warning
When using direct `field` property, keep in mind that the model will be `null` during loading
:::

To display the loading state of the component, when the data is being fetched, you can use the `loading` slot.

```vue
<template>
  [...]
    <input-field :model="album" field-name="title">
      <template v-slot:loading>
        <span>Loading title...</span>
      </template>
    </input-field>
  [...]
</template>
```

## FieldLabel

In case you just want to render the label of the field you can use the `FieldLabel` component which will resolve the async field label. 
You can use your model instance or the static model class

```vue
<template>
  [...]
    <!-- with model static class -->
    <field-label :model="Album" field-name="title"/>
    <!-- or directly with field -->
    <field-label :field="Album.getField('title')"/>

    <!-- with model instance -->
    <field-label :model="album" field-name="title"/>
    <!-- or directly with field -->
    <field-label :field="album.getField('title')"/>
  [...]
</template>
```