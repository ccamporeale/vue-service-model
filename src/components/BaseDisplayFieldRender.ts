import AsyncComputed from 'vue-async-computed'
import Vue, { VNode, CreateElement } from 'vue'
import { Field } from '../fields/Field'
import cu from '../utils/common'

Vue.use(AsyncComputed)

export interface ComponentData {
  // AsyncComputed
  renderData?: any
}

export default Vue.extend({
  name: 'BaseDisplayFieldRender',
  inheritAttrs: false,

  props: {
    field: {
      type: Field,
      required: true
    }
  },

  data: (): ComponentData => ({}),

  computed: {
    hasResolvedRenderData () {
      return this.renderData !== cu.NO_VALUE
    }
  },

  asyncComputed: {
    renderData: {
      default: cu.NO_VALUE,
      get () {
        const field = this.field as Field
        return field.prepareDisplayRender()
      }
    }
  },

  methods: {
    renderField (h: CreateElement): VNode {
      return this.field.displayRender(h, this.renderData)
    }
  },

  render (h: CreateElement): VNode {
    if (this.hasResolvedRenderData) {
      return this.renderField(h)
    } else {
      return undefined as any
    }
  }
})
