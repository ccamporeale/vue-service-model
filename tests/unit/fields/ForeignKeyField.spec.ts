import Vue, { CreateElement } from 'vue'
import { mount } from '@vue/test-utils'
import { BaseModel } from '@/models/BaseModel'
import { ServiceModel } from '@/models/ServiceModel'
import { CharField } from '@/fields/CharField'
import { ForeignKeyField } from '@/fields/ForeignKeyField'
import { InvalidFieldOptionsException, RequiredFieldOptionsException } from '@/exceptions/FieldExceptions'
import DisplayField from '@/components/DisplayField'
import { waitRender } from '../../testUtils'
import { ComponentModule } from '@/types/components'
import BaseDisplayFieldRender from '@/components/BaseDisplayFieldRender'

describe('fields/ForeignKeyField', () => {
  class User extends ServiceModel {
    static fieldsDef = {
      name: new (class extends CharField {
        get displayComponent (): Promise<ComponentModule> {
          return Promise.resolve({
            default: Vue.extend({
              props: {
                field: {
                  type: Object,
                  required: true
                }
              },
              render (h: CreateElement) {
                return h('div', this.field.data.name)
              }
            })
          })
        }
      })()
    }
  }

  const user = new User({ name: 'User name' })

  class TestModel extends BaseModel {
    static fieldsDef = {
      user: new ForeignKeyField({
        options: {
          model: User,
          fieldName: 'name'
        }
      })
    }
  }

  describe('get value', () => {
    it('should fetch foreign instance', async () => {
      const mockObjectsDetail = jest.spyOn(User.objects, 'detail').mockImplementation(async () => user)

      const userId = 0
      const model = new TestModel({ user: userId })
      const value = await model.val.user

      expect(value).toBe(user)

      expect(mockObjectsDetail).toBeCalledTimes(1)
      expect(mockObjectsDetail.mock.calls[0]).toEqual([userId])
      mockObjectsDetail.mockRestore()
    })

    it('should not fetch on null value', async () => {
      const mockObjectsDetail = jest.spyOn(User.objects, 'detail').mockImplementation(async () => user)

      const model = new TestModel()
      const value = await model.val.user

      expect(value).toBeNull()

      expect(mockObjectsDetail).toBeCalledTimes(0)
      mockObjectsDetail.mockRestore()
    })
  })

  describe('options', () => {
    it('should return valid options', async () => {
      const field = TestModel.getField('user')
      const options = await field.options
      expect(options).toEqual({
        model: User,
        fieldName: 'name'
      })
    })

    it('should throw RequiredFieldOptionsException for missing options.model', async () => {
      const field = new ForeignKeyField({ options: {} })
      await expect(field.options).rejects.toBeInstanceOf(RequiredFieldOptionsException)
    })

    it('should throw RequiredFieldOptionsException for missing options.fieldName', async () => {
      const field = new ForeignKeyField({ options: { model: User } })
      await expect(field.options).rejects.toBeInstanceOf(RequiredFieldOptionsException)
    })

    it('should throw InvalidFieldOptionsException for missing options.fieldName', async () => {
      const field = new ForeignKeyField({ options: { model: {}, fieldName: 'name' } })
      await expect(field.options).rejects.toBeInstanceOf(InvalidFieldOptionsException)
    })
  })

  describe('prepareDisplayRender', () => {
    it('should return foreign field instance', async () => {
      const model = new TestModel({ user: 8 })
      const field = model.getField('user')
      const mockObjectsDetail = jest.spyOn(User.objects, 'detail').mockImplementation(async () => user)

      const renderData = await field.prepareDisplayRender()

      expect(Object.keys(renderData)).toHaveLength(2)

      expect(renderData.field).toBe(user.getField('name'))
      expect(renderData.displayField).toBe(DisplayField)

      expect(mockObjectsDetail).toBeCalledTimes(1)
      mockObjectsDetail.mockRestore()
    })

    it('should return no field data', async () => {
      const model = new TestModel()
      const field = model.getField('user')

      const renderData = await field.prepareDisplayRender()

      expect(renderData).toEqual({
        field: null,
        displayField: null
      })
    })
  })

  describe('displayRender', () => {
    it('should render correct value', async () => {
      const model = new TestModel({ user: 9 })
      const mockObjectsDetail = jest.spyOn(User.objects, 'detail').mockImplementation(
        async pk => pk === 9 ? user : new User({ name: 'User 10' }))
      const field = model.getField('user')

      const wrapper = mount(BaseDisplayFieldRender, {
        propsData: { field: field }
      })

      await waitRender.DisplayField(wrapper)
      await waitRender.DisplayField(wrapper)
      expect(wrapper.html()).toBe('')

      expect(mockObjectsDetail).toBeCalledTimes(1)
      expect(mockObjectsDetail.mock.calls[0]).toEqual([9])

      await waitRender.DisplayField(wrapper)
      expect(wrapper.html()).toMatchSnapshot()

      model.val.user = 10
      expect(model.data).toEqual({ user: 10 })

      await waitRender.DisplayFieldUpdate(wrapper)
      await waitRender.DisplayFieldUpdate(wrapper)

      expect(mockObjectsDetail).toBeCalledTimes(2)
      expect(mockObjectsDetail.mock.calls[1]).toEqual([10])

      await waitRender.DisplayFieldUpdate(wrapper)

      expect(wrapper.html()).toMatchSnapshot()

      mockObjectsDetail.mockRestore()
    })
  })
})
