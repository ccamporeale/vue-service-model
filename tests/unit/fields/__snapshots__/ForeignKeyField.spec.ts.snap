// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fields/ForeignKeyField displayRender should render correct value 1`] = `<div>User name</div>`;

exports[`fields/ForeignKeyField displayRender should render correct value 2`] = `<div>User 10</div>`;
